packages/koa-shopify-auth/src/auth/test/index.test.ts

```diff
@@ -0,0 +1,158 @@
+  scopes: ['write_orders, write_products'],
+  accessMode: 'offline',
+};
+
+describe('Index', () => {
+  describe('with the /auth path', () => {
+    describe('with no test cookie', () => {
+      it('redirects to enable cookies', async () => {
+        const shopifyAuth = createShopifyAuth(baseConfig);
+        const ctx = createMockContext({
+          url: `https://${baseUrl}`,
+        });
+        const nextFunction = () => {};
+        const mockTopLevelRedirect = jest.fn();
+        createTopLevelRedirect.mockImplementation(mockTopLevelRedirect);
+
+        await shopifyAuth(ctx, nextFunction);
+
+        expect(createTopLevelRedirect).toBeCalledWith('/auth/enable_cookies');
+        expect(mockTopLevelRedirect).toBeCalledWith(ctx);
+      });
+    });
+
+    describe('with a test cookie but not top-level cookie', () => {
+      it('redirects to /auth/inline at the top-level', async () => {
+        const shopifyAuth = createShopifyAuth(baseConfig);
+        const ctx = createMockContext({
+          url: `https://${baseUrl}`,
+          cookies: {'shopifyTestCookie': '1'},
+        });
+        const nextFunction = () => {};
+        const mockTopLevelRedirect = jest.fn();
+        createTopLevelRedirect.mockImplementation(mockTopLevelRedirect);
+
+        await shopifyAuth(ctx, nextFunction);
+
+        expect(createTopLevelRedirect).toBeCalledWith('/auth/inline');
+        expect(mockTopLevelRedirect).toBeCalledWith(ctx);
+      });
+
+      it('sets the top-level cookie', async () => {
+        const shopifyAuth = createShopifyAuth(baseConfig);
+        const ctx = createMockContext({
+          url: `https://${baseUrl}`,
+          cookies: {'shopifyTestCookie': '1'},
+        });
+        const nextFunction = () => {};
+
+        await shopifyAuth(ctx, nextFunction);
+
+        expect(ctx.cookies.set).toBeCalledWith('shopifyTopLevelOAuth', '1');
+      });
+    });
+
+    describe('with a test cookie and a top-level cookie', () => {
+      it('performs inline oauth', async () => {
+        const shopifyAuth = createShopifyAuth(baseConfig);
+        const ctx = createMockContext({
+          url: `https://${baseUrl}`,
+          cookies: {'shopifyTestCookie': '1', 'shopifyTopLevelOAuth': '1'},
+        });
+        const nextFunction = () => {};
+
+        await shopifyAuth(ctx, nextFunction);
+
+        expect(mockOAuthStart).toBeCalledWith(ctx);
+      });
+
+      it('clears the top-level cookie', async () => {
+        const shopifyAuth = createShopifyAuth(baseConfig);
+        const ctx = createMockContext({
+          url: `https://${baseUrl}`,
+          cookies: {'shopifyTestCookie': '1', 'shopifyTopLevelOAuth': '1'},
+        });
+        const nextFunction = () => {};
+
+        await shopifyAuth(ctx, nextFunction);
+
+        expect(ctx.cookies.set).toBeCalledWith('shopifyTopLevelOAuth');
+      });
+    });
+  });
+
+  describe('with the /auth/inline path', () => {
+    it('performs inline oauth', async () => {
```

Comment:
I'm curious how you found working with `createMockContext`, since it hasn't gotten much use outside of quilt yet.

I think we probably also want some E2E tests on this at some point, but this changeset is already quite large so don't worry yet.